pkg/appgw/frontend_listeners.go

```diff
@@ -1,16 +1,97 @@
 package appgw
 
-import "k8s.io/api/extensions/v1beta1"
+import (
+	"strings"
 
-func (builder *appGwConfigBuilder) getListenerConfigs(ingressList []*v1beta1.Ingress) map[frontendListenerIdentifier]*frontendListenerAzureConfig {
-	httpListenersAzureConfigMap := make(map[frontendListenerIdentifier]*frontendListenerAzureConfig)
+	n "github.com/Azure/azure-sdk-for-go/services/network/mgmt/2018-12-01/network"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/golang/glog"
+	"k8s.io/api/extensions/v1beta1"
+)
 
+// getFrontendListeners constructs the unique set of App Gateway HTTP listeners across all ingresses.
+func (builder *appGwConfigBuilder) getFrontendListeners(ingressList []*v1beta1.Ingress) (*[]n.ApplicationGatewayHTTPListener, map[frontendListenerIdentifier]*n.ApplicationGatewayHTTPListener) {
+	// TODO(draychev): this is for compatibility w/ RequestRoutingRules and should be removed ASAP
+	legacyMap := make(map[frontendListenerIdentifier]*n.ApplicationGatewayHTTPListener)
+
+	var httpListeners []n.ApplicationGatewayHTTPListener
+
+	for listener, config := range builder.getListenerConfigs(ingressList) {
+		httpListener := builder.newHTTPListener(listener, config.Protocol)
+		if config.Protocol == n.HTTPS {
+			sslCertificateID := builder.appGwIdentifier.sslCertificateID(config.Secret.secretFullName())
+			httpListener.SslCertificate = resourceRef(sslCertificateID)
+		}
+		httpListeners = append(httpListeners, httpListener)
+		legacyMap[listener] = &httpListener
+	}
+
+	// TODO(draychev): The second map we return is for compatibility w/ RequestRoutingRules and should be removed ASAP
+	return &httpListeners, legacyMap
+}
+
+// getListenerConfigs creates an intermediary representation of the listener configs based on the passed list of ingresses
+func (builder *appGwConfigBuilder) getListenerConfigs(ingressList []*v1beta1.Ingress) map[frontendListenerIdentifier]frontendListenerAzureConfig {
+	allListeners := make(map[frontendListenerIdentifier]frontendListenerAzureConfig)
 	for _, ingress := range ingressList {
-		_, _, configMap := builder.processIngressRules(ingress)
-		for k, v := range configMap {
-			httpListenersAzureConfigMap[k] = v
+		_, azListenerConfigs := builder.processIngressRules(ingress)
+		for listenerID, azConfig := range azListenerConfigs {
+			allListeners[listenerID] = azConfig
```

Comment:
We will be overwriting existing keys here. If 2 ingress have no hostname and same port, ("", 80)
This means the ingress config is conflicting.
Let's discuss.

Reply:
Agreed - needs refinement in how we handle duplicates. Behavior was the same with the UnorderedSets - one copy was kept at any time.  Will discuss and find a better way to handle this asap. (taking a note to come back to this)